### PR TITLE
ENH: Addressing #1393

### PR DESCRIPTION
**Pull request by mortonjt**
Fixing _check_orthogonality
Addressing #1393

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,7 +9,7 @@
 ### Backward-incompatible changes [experimental]
 
 ### Bug fixes
-
+* Improving useability for `ilr` and `ilr_inv` for 2D compositions.  ([#1393](https://github.com/biocore/scikit-bio/issues/1393))
```
Comment (jairideout) [inline CHANGELOG.md]: Can you expand the description of the bug? From this description, I can't tell what input(s) caused the bug and what the buggy behavior was (was it a silent bug producing incorrect results, or did it error noisily?). It appears the `basis` parameter didn't work with 1-D arrays for `ilr` and `ilr_inv`, but I'm unsure about the `clr_inv` transformation to `basis` that you added to `ilr_inv`. If there's more than one issue with `basis`, please describe that here.

The markdown heading below this section doesn't render correctly, please add a blank line after this bullet point.

Please also indicate the full public import path for these functions, i.e. `skbio.stats.composition.ilr` and `skbio.stats.composition.ilr_inv`.

**Comment by jairideout** [comment]
Thanks for the fix @mortonjt! I just had one comment related to informing users about the bug (changelog).

File: CHANGELOG.md [hunk 2]
```diff
@@ -9,6 +9,7 @@
 ### Backward-incompatible changes [experimental]
 
 ### Bug fixes
+* Modified `skbio.stats.composition.ilr` and `skbio.stats.composition.ilr_inv` to suppress errors when inputing a 1D basis.  ([#1393](https://github.com/biocore/scikit-bio/issues/1393))
```
Comment (jairideout) [inline CHANGELOG.md]: Thanks for expanding, this looks great! Just to confirm, what is the rationale for changing `ilr_inv` handling of `basis` from:

```python
_check_orthogonality(basis)
```

to:

```python
_check_orthogonality(clr_inv(basis))
```
It seems unrelated to handling 1-D basis. Is this a separate bug? If so please add that to the changelog.

Comment (mortonjt) [inline CHANGELOG.md]: Sorry about that.  Yeah, that was a separate bug.  Added.

File: CHANGELOG.md [hunk 3]
```diff
@@ -9,6 +9,7 @@
 ### Backward-incompatible changes [experimental]
 
 ### Bug fixes
+* Modified `skbio.stats.composition.ilr` and `skbio.stats.composition.ilr_inv` to suppress errors when inputing a 1D basis.  ([#1393](https://github.com/biocore/scikit-bio/issues/1393)).  Also fixing how the basis is handled in `skbio.stats.composition.ilr_inv`.
```
Comment (jairideout) [inline CHANGELOG.md]: Please update the second bug description to include what the previous behavior was (was it a silent bug that produced incorrect results, did it fail noisily, or something else?) and what the new behavior is.

In general, when a bug fix is submitted it is essential that the changelog entry includes these details as this is how we notify our users. This is analogous to QIIME 1's blog posts describing bugs (they include the same information for users).

Comment (jairideout) [inline CHANGELOG.md]: Also, can you squash your commits when that's addressed?

Comment (mortonjt) [inline CHANGELOG.md]: Alrighty.  Here's a run down of what happened.

- `_check_orthogonality` only accepts a basis that is within the simplex.  The basis passed into `ilr_inv` lives only in `ilr` space, so in order to make sure that the basis is correct (i.e. orthogonal), it needs to be transformed into the simplex space to match `_check_orthogonality`
- So basically, you need to be consistent between these two spaces (`ilr` space, and simplex space).  And I just fixed this mistake to enforce this consistency.  I didn't catch this before, because we didn't have a test case for passing in a basis for `ilr_inv`.

We now have that fix.  Let me know if there is anything else that I need to clarify.

File: CHANGELOG.md [hunk 4]
```diff
@@ -9,6 +9,7 @@
 ### Backward-incompatible changes [experimental]
 
 ### Bug fixes
+* Modified `skbio.stats.composition.ilr` and `skbio.stats.composition.ilr_inv` to suppress errors when inputing a 1D basis.  ([#1393](https://github.com/biocore/scikit-bio/issues/1393)).  Also fixing how the basis is handled in `skbio.stats.composition.ilr_inv`, the basis is now transformed back into the simplex space before checking to see if the basis is orthogonal.
```
Comment (ebolyen) [inline CHANGELOG.md]: Instead of "suppress errors" how about "coerces rank to 2D" or something similar?

`Also ...` should be a second bullet with past tense. Examples would be good.

**Comment by mortonjt** [comment]
I don't think this is a good use of time to be spending on perfecting a 2 sentence bug report.

If there is still misunderstanding on exactly what this functionality is doing, let's schedule a call after Scipy.  Otherwise, I recommend merging this fix as is.

**Comment by mortonjt** [comment]
Had a discussion offline with @ebolyen about the functionality that has been modified.

Here's exactly what is going on with the `ilr_inv` method.

There is no bug in the existing interface.  But there is some potential confusion for user concerning the input basis for `ilr` and `ilr_inv`.   The `ilr` method takes in a `basis` that is `D-1 x D` matrix (i.e. proportions by balances).  The `ilr_inv` method takes in a `basis` that is `D x D-1` matrix (i.e. balances by proportions).  This can be confusing for users, because this would require that the basis needs to be also transformed when applying `ilr` and `ilr_inv`.

In other words, if you wanted to run the following code
```
>>> from skbio.stats.composition import ilr, ilr_inv, clr
>>> import numpy as np
>>> basis = np.array([[0.80442968, 0.19557032]])
>>> table = np.array([[np.log(1/10)*np.sqrt(1/2),
...                            np.log(1.14141414 / 9.90909091)*np.sqrt(1/2),
...                            np.log(1.28282828 / 9.81818182)*np.sqrt(1/2),
...                            np.log(1.42424242 / 9.72727273)*np.sqrt(1/2),
...                            np.log(1.56565657 / 9.63636364)*np.sqrt(1/2)]]).T
>>> res = ilr(ilr_inv(table, basis=basis), basis=basis)  
>>> res
array([-0.70097524, -0.65792874, -0.61956576, -0.58489877, -0.55322134])
>>> table
array([-1.62817353, -1.52818832, -1.4390816 , -1.35855968, -1.2849817 ])      
```

`res` and `table` are different, even though they should be the same, since `ilr` and `ilr_inv` are inverses of each other.

In this PR, I have made this easier to run.  More specifically, I am making sure that `ilr` and `ilr_inv` can take in the exact same basis.  And that basis lives in the Aitchison simplex.  

In other words. The functions have the following signatures

```
ilr (x, e) :: (P, P) -> B
ilr_inv(y, e) :: (B, P) -> P
```

Where `x` is a `n x D` matrix of proportions (`n` samples and `D` proportions),
`y` is a `n x D-1` matrix of balances (`n` samples and `D-1` balances),
`e` is a `D-1 x D` matrix representing a basis of proportions (`D-1` vectors of `D` proportions),
`P` represents the space of proportions (i.e. Aitchison simplex), and `B` represents the space of balances (i.e. ilr space).  

An additional test under`test_ilr_basis_isomorphism` explicitly ensures that `ilr` and `ilr_inv` are inverses of each other if the same basis is applied to both functions.

Hopefully this makes this function a little more clear, and easier to use.  Let me know if you have any more questions about this. Thanks!

**Comment by mortonjt** [comment]
Also note that we won't be worrying about the issue brought up in #1393.  This method should not take in 1D matrices, since the `basis` parameter should be a 2D matrix.  A ValueError is now raised to catch that.

File: CHANGELOG.md [hunk 5]
```diff
@@ -9,6 +9,7 @@
 ### Backward-incompatible changes [experimental]
 
 ### Bug fixes
+* Modifying basis handling in `skbio.stats.composition.ilr_inv` prior to checking for orthogonality.  Now the basis is strictly assumed to be in the Aitchison simplex. ([#1393](https://github.com/biocore/scikit-bio/issues/1393))
```
Comment (ebolyen) [inline CHANGELOG.md]: Won't have a chance to really review until the morning, but real quick comment:
This should be in the section just above (e.g. in `Backward-incompatible changes [experimental]`) as code that was already accounting for the inverse basis will need to be adjusted.
Thanks!

**Comment by ebolyen** [comment]
Thanks for the explanation @mortonjt!

File: skbio/stats/composition.py [hunk 6]
```diff
@@ -548,6 +551,12 @@ def ilr(mat, basis=None, check=True):
...
 
+    Notes
+    -----
+    If the `basis` parameter is specified, it is expected to be a basis in the
+    Aitchison simplex.  If there are `D-1` elements specified in `mat`, then
+    the dimensions of the basis needs be `D-1 x D`, where rows represent
+    basis vectors, and the columns represent proportions.
     """
     mat = closure(mat)
     if basis is None:
```
Comment (ebolyen) [inline skbio/stats/composition.py]: Commenting here because GitHub...

Just after line 563 we should perform the same check for 2D as in `ilr_inv`.

File: CHANGELOG.md (hunk 5 above)
Comment (ebolyen) [inline CHANGELOG.md]: Also the issue number/link is no longer relevant to this change, so you can just drop it entirely.

**Comment by ebolyen** [comment]
@mortonjt Done reviewing, two little things and then this should be good to go.

**Comment by mortonjt** [comment]
Addressed both comments.  Thanks @ebolyen !

**Comment by ebolyen** [comment]
Thanks @mortonjt, this should be ready for merge when tests pass.

**Comment by ebolyen** [comment]
@mortonjt looks like we have  a PEP8 error:
```
skbio/stats/composition.py:574:1: E302 expected 2 blank lines, found 1
```

**Comment by ebolyen** [comment]
Thanks @mortonjt!

**Comment by mortonjt** [comment]
🎆 Thanks!!
